src/main/java/org/opentestsystem/shared/contentspecid/ContentSpecIdBuilder.java

```diff
@@ -0,0 +1,241 @@
+
+        // Derived values
+        DomainCode derivedDomain = deriveDomain(subject, grade, convertedClaim, convertedTarget);
+        boolean isMathClaim2Thru4 = (subject == MATH && convertedClaim != ContentSpecClaim.C1);
+        String convertedDomain = convertDomain(domain, derivedDomain, isMathClaim2Thru4, grade, ccss);
+
+        Emphasis emphasis = deriveEmphasis(this.subject, convertedClaim,
+                grade, convertedTarget, ccss);
+        validateEmphasis(this.emphasis, emphasis);
+
+        String contentCategory = deriveContentCategory(this.subject, convertedClaim, emphasis, convertedDomain);
+        validateContentCategory(this.contentCategory, contentCategory);
+        int targetSet = deriveTargetSet(this.subject, grade, convertedClaim);
+
+        return new ContentSpecId(
+                this.subject,
+                this.grade,
+                convertedClaim,
+                convertedTarget,
+                this.ccss,
+                convertedDomain,
+                emphasis,
+                contentCategory,
+                targetSet);
+    }
+
+    private String convertTarget(String target) {
+        String convertedTarget = trimToEmpty(target);
+        if (convertedTarget.toUpperCase().startsWith("T") && target.length() > 1) {
+            convertedTarget = convertedTarget.substring(1);
+        }
+        return convertedTarget;
+    }
+
+    private ContentSpecClaim convertClaim(String claimString) throws IdParseException {
+        if (isBlank(claim)) {
+            return ContentSpecClaim.UNSPECIFIED;
+        }
+
+        ContentSpecClaim claim = ContentSpecClaim.fromString(claimString);
+        if (claim != ContentSpecClaim.UNSPECIFIED) {
+            return claim;
+        }
+
+        throw new IdParseException(ErrorSeverity.INVALID,
+                String.format("Expected claim but found '%s'.", claimString));
+    }
+
+    private String convertDomain(String domain, DomainCode derivedDomain, boolean mathClaim2thru4,
+                                 ContentSpecGrade grade, String ccss)
+            throws IdParseException {
+
+        if (StringUtils.isBlank(domain) || domain.equals(NOT_APPLICABLE)) {
+            return derivedDomain.getValue();
+        }
+
+        String convertedDomain = domainHandler.stripDashedDomain(domain, grade, ccss);
+        // Math claims 2 to 4 will not have a domain, so accept the set one.
+        // TODO: In some cases this can be a domain followed by extraneous characters. Why?
+        if (mathClaim2thru4) {
+            return convertedDomain;
+        }
+
+        if (derivedDomain == DomainCode.UNK) {
+            return convertedDomain;
+        }
+
+        // Otherwise, check that the value is what we expect it to be.
+        if (!(convertedDomain.equals(derivedDomain.getValue())
+                || convertedDomain.equals(derivedDomain.getLegacyValue()))) {
+            throw new IdParseException(ErrorSeverity.CORRECTED,
```

Comment:
Can we change the name of IdParseException into ParseException?  Or ValidationException?  It seems like this Exception type is used more than just for `id` errors.

Reply:
I was trying to avoid the name conflict with java.text.ParseException, which I'd normally use for a parser, but it doesn't have an error severity. The c-sharp library handles errors a bit differently, but does distinguish between parse errors caused by going from string to pojo (pocso?) and validation errors caused by going from pojo to string. I'd be fine just going with ValidationException for both. What do you think?

Reply:
Do we handle parse exceptions any differently than Validation Exceptions?  If not, I think it would be fine to combine.  I was focused on the `id` in the name itself.  I now see that `id` is because we're parsing ids.  Id is just an overloaded term for me in software.